### PR TITLE
subworlds

### LINKED ISSUE #144: Sub-worlds
A crazy idea that just came up in IRC is to have certain cells function as entrances and exits to other sub-worlds.  e.g. there could be caves, buildings, ...  This seems like it would be fun and could be an easy way to add new stuff to the world without taking up a ton of space.

Concretely I don't think this would be all that hard to add (famous last words, right?).  A meta-world will now consist of a collection of worlds, each with an integer ID.  A location now consists of a pair of (world id, coordinates).  An entrance to another area just stores the location that it links to.

### PR DESCRIPTION
**Pull request by kostmo**
Closes #144.

This builds upon portals support (#1356)

# Demo

    scripts/play.sh --scenario data/scenarios/Testing/144-subworlds/subworld-mapped-robots.yaml --autoplay --speed 2

[asciicast](https://asciinema.org/a/vC13dW8M1S8t2b1J4XkW80U1q)

# Future work
* Augment portal definitions with an optional "relative orientation" attribute, that can turn the player around when passing through the portal (#1379)
* Specify whether portal performs instant transportation or whether `move down` is required (#1368)

### CODE REVIEW

**Comment by kostmo** [comment]
@byorgey have a look at your leisure.  There's a bit more work to be done, but I'm excited to say that this "works" and passes the test suite.

File: src/Swarm/Game/World.hs [hunk 1]
```diff
@@ -31,7 +32,9 @@ module Swarm.Game.World (
   -- ** World functions
   newWorld,
   emptyWorld,
+  lookupCosmosTerrain,
```
Comment (byorgey) [inline src/Swarm/Game/World.hs:35]: Is it `cosmo` or `cosmos`?

File: src/Swarm/Game/World.hs [hunk 2]
```diff
@@ -225,31 +241,43 @@ lookupTerrain i (World f t _) =
   ((U.! tileOffset i) . fst <$> M.lookup (tileCoords i) t)
     ? fst (runWF f i)
 
--- | A stateful variant of 'lookupTerrain', which first loads the tile
+-- | A stateful variant of "lookupTerrain", which first loads the tile
```
Comment (byorgey) [inline src/Swarm/Game/World.hs:244]: Note the single quotes are semantically meaningful --- they cause Haddock to turn the quoted thing into a link to the named function or type.

File: src/Swarm/Game/Scenario/Topography/Navigation/Portal.hs [hunk 3]
```diff
@@ -97,24 +153,109 @@ validateNavigation upperLeft unmergedWaypoints portalDefs = do
...
+      else Positive pair
+
+  tuplify = both (view subworld) &&& both (view planar)
+
+  nest ::
+    Signed (b, a) ->
+    (b, Signed a)
+  nest = \case
+    Positive x -> fmap Positive x
+    Negative x -> fmap Negative x
```
Comment (kostmo) [inline src/Swarm/Game/Scenario/Topography/Navigation/Portal.hs:252]: I thought `sequenceA` would work here instead, but the compiler doesn't like it.

File: src/Swarm/Game/State.hs [hunk 4]
```diff
@@ -686,14 +693,22 @@ messageNotifications = to getNotif
...
+-- | Reconciles the possibilities of log messages being
+-- omnipresent and robots being in different worlds
+messageIsFromNearby :: Cosmo Location -> LogEntry -> Bool
+messageIsFromNearby l e = case e ^. leLocation of
+  Omnipresent -> True
+  Located x -> f x
+ where
+  f logLoc = case cosmoMeasure manhattan l logLoc of
+    InfinitelyFar -> False
+    Measurable x -> x <= hearingDistance
```
Comment (kostmo) [inline src/Swarm/Game/State.hs:705]: I clarified the semantics of log message proximity.  The code had been suffering from ["Maybe blindness"](https://www.reddit.com/r/haskell/comments/8eecyu/comment/dxuqb33/?utm_source=share&utm_medium=web2x&context=3) IMO, so I created two new isomorphic types:
* `DistanceMeasure`
* `LogLocation`

File: src/Swarm/Game/Universe.hs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import Data.Function (on)
+import Data.Int (Int32)
+import Data.Text (Text)
+import Data.Yaml (FromJSON, ToJSON, Value (Object), parseJSON, withText, (.:))
+import GHC.Generics (Generic)
+import Linear (V2 (..))
+import Swarm.Game.Location
+
+data SubworldName = DefaultRootSubworld | SubworldName Text
+  deriving (Show, Eq, Ord, Generic, ToJSON)
```
Comment (kostmo) [inline src/Swarm/Game/Universe.hs:18]: This `DefaultRootSubworld` constructor worked out nicely.  Previously I was using an "in-band" constant name `"root"` as a default (i.e. when a subworld name is not provided in the scenario file), which became problematic.

File: data/scenarios/Testing/_Validation/144-subworld-uniqueness-enforcement-default.yaml [hunk 6]
```diff
@@ -0,0 +1,87 @@
+version: 1
+name: Subworld uniqueness (default name)
+description: |
+  Has two unnamed subworlds, which fail uniquenss
```
Comment (kostmo) [inline data/scenarios/Testing/_Validation/144-subworld-uniqueness-enforcement-default.yaml:4]: Typo

File: src/Swarm/Game/Scenario/Topography/Navigation/Portal.hs [hunk 7]
```diff
@@ -97,24 +153,109 @@ validateNavigation upperLeft unmergedWaypoints portalDefs = do
...
+-- For all of the portals between Subworlds A and B:
+-- * The coordinates of all \"consistent\" portal locations in Subworld A
+--   are subtracted from the corresponding coordinates in Subworld B. It
+--   does not matter which are exits vs. entrances.
+-- * The resulting \"vector\" from every pair must be equal.
+ensureSpatialConsistency ::
+  MonadFail m =>
+  [(Cosmo Location, AnnotatedDestination Location)] ->
+  m ()
+ensureSpatialConsistency xs =
```
Comment (kostmo) [inline src/Swarm/Game/Scenario/Topography/Navigation/Portal.hs:226]: Just realized this implementation is not yet correct; I had based it on a test case that is "mirrored", not consistent.

File: src/Swarm/Game/Scenario/Topography/WorldDescription.hs [hunk 8]
```diff
@@ -32,36 +37,49 @@ data PWorldDescription e = WorldDescription
...
   deriving (Eq, Show)
 
 type WorldDescription = PWorldDescription Entity
 
-instance FromJSONE (EntityMap, RobotMap) WorldDescription where
+instance FromJSONE (InheritedStructureDefs, (EntityMap, RobotMap)) WorldDescription where
   parseJSONE = withObjectE "world description" $ \v -> do
-    pal <- v ..:? "palette" ..!= WorldPalette mempty
-    structureDefs <- v ..:? "structures" ..!= []
+    (scnenarioLevelStructureDefs, (em, rm)) <- getE
```
Comment (kostmo) [inline src/Swarm/Game/Scenario/Topography/WorldDescription.hs:49]: Typo

File: src/Swarm/Game/Scenario/Topography/Navigation/Portal.hs (hunk 7 above)
Comment (kostmo) [reply src/Swarm/Game/Scenario/Topography/Navigation/Portal.hs:226]: Scratch that, the implementation is correct, after all.  The test case was "mirrored", but it was "consistently mirrored", so did exercise the validation logic properly.

File: src/Swarm/Game/Scenario/Topography/Navigation/Portal.hs (hunk 3 above)
Comment (kostmo) [reply src/Swarm/Game/Scenario/Topography/Navigation/Portal.hs]: Resolved this question in b1a771e.

**Comment by kostmo** [comment]
This is ready for a final look, @byorgey.

**Comment by byorgey** [comment]
@kostmo can you fix (or remove) the link to the asciinema preview?

**Comment by kostmo** [comment]
> @kostmo can you fix (or remove) the link to the asciinema preview?

Fixed.  I wonder if there is a way to keep these recordings with longer retention?

**Comment by byorgey** [comment]
> Fixed. I wonder if there is a way to keep these recordings with longer retention?

I think you have to sign up for an account and upload it to your account, then you can keep them forever.  It looks like anonymous ones are deleted after a week.

**Comment by byorgey** [comment]
@kostmo Also, could you link to an issue for the other "future work" bullet?  I wasn't sure if one already exists or if we need to create one.

File: src/Swarm/Game/Scenario.hs [hunk 9]
```diff
@@ -123,6 +133,37 @@ instance FromJSONE EntityMap Scenario where
...
+        rootWorld <- v ..: "world"
+        subworlds <- v ..:? "subworlds" ..!= []
+        return $ rootWorld :| subworlds
+
+      let worldsByName = binTuples $ NE.toList $ NE.map (worldName &&& id) allWorlds
+          dupedNames = M.keys $ M.filter ((> 1) . length) worldsByName
+      unless (null dupedNames)
+        . fail
+        . T.unpack
+        $ T.unwords
```
Comment (byorgey) [inline src/Swarm/Game/Scenario.hs:148]: Use `failT`.

File: src/Swarm/Game/Scenario.hs [hunk 10]
```diff
@@ -92,7 +101,8 @@ data Scenario = Scenario
   , _scenarioEntities :: EntityMap
   , _scenarioRecipes :: [Recipe Entity]
   , _scenarioKnown :: [Text]
-  , _scenarioWorld :: WorldDescription
+  , _scenarioWorlds :: NonEmpty WorldDescription
```
Comment (byorgey) [inline src/Swarm/Game/Scenario.hs:104]: If the root world must always come first, let's document that invariant somewhere.

**Review by byorgey — COMMENTED** [review]
This is looking great.  Only one or two things that I think absolutely must be addressed before merging (e.g. the thing with `sequenceTuple`) and then a bunch of style/documentation suggestions.

I will have my hands full figuring out a way to support all this cool functionality in the new world description DSL, which is a great problem to have. :smile:

File: data/scenarios/Testing/144-subworlds/subworld-located-robots.yaml [hunk 11]
```diff
@@ -0,0 +1,116 @@
+version: 1
+name: Subworld robots (explicit location)
+description: |
+  Demonstrate that system robots can be placed in any subworld.
```
[the remark this answers is not in the document]
Comment (kostmo) [reply data/scenarios/Testing/144-subworlds/subworld-located-robots.yaml:4]: Created #1381

File: src/Swarm/Util.hs [hunk 12]
```diff
@@ -189,6 +191,25 @@ findDup = go S.empty
...
+
+-- | This function has a lamentable basis.
+-- The 'sequenceA' function requires an 'Applicative' instance
+-- for the inner 'Functor'. However, the 'Applicative' instance
+-- of @(,)@ (the two-element tuple) requires a 'Monoid' instance
+-- for the first element!
+-- See: https://hackage.haskell.org/package/base-4.18.0.0/docs/src/GHC.Base.html#line-523
+--
+-- The 'sequenceA' operation does not affect the first element
+-- of the tuple, so it shouldn't matter whether it has a 'Monoid' instance!
```
[the remark this answers is not in the document]
Comment (kostmo) [reply src/Swarm/Util.hs]: Fixed in 659c7a0.

[the remark this answers is not in the document]
Comment (kostmo) [reply src/Swarm/Util.hs]: Regarding my commentary in that commit, come to think of it, I suppose that a datatype `Foo a` that consists of exclusively unary data constructors would be isomorphic to a type `(Foo, a)` where `Foo` consists exclusively of nullary constructors.

E.g.:
```
Foo a = C1 a | C2 a | C3 a
```
vs.
```
Foo = C1 | C2 | C3
FooIso a = (Foo, a)
```

[the remark this answers is not in the document]
Comment (kostmo) [reply src/Swarm/Util.hs]: Ah, but the point is moot, as the same workaround is required (demo in 46e14ad2).

File: src/Swarm/TUI/Model/StateUpdate.hs [hunk 13]
```diff
@@ -207,8 +208,7 @@ scenarioToUIState isAutoplaying siPair@(scenario, _) gs u = do
  where
   entityList = EU.getEntitiesForList $ gs ^. entityMap
 
-  myWorld = scenario ^. scenarioWorld
-  (isEmptyArea, newBounds) = EU.getEditingBounds myWorld
+  (isEmptyArea, newBounds) = EU.getEditingBounds $ NE.head $ scenario ^. scenarioWorlds
```
[the remark this answers is not in the document]
Comment (kostmo) [reply src/Swarm/TUI/Model/StateUpdate.hs:211]: Now documented.

**Comment by kostmo** [comment]
All comments now addressed, @byorgey.

File: src/Swarm/Game/Scenario/Topography/Navigation/Portal.hs [hunk 14]
```diff
@@ -1,38 +1,65 @@
...
--- the reserved name \"root\".
-newtype SubworldName = SubworldName Text
-  deriving (Show, Eq, Ord, Generic, FromJSON)
+type WaypointMap = M.Map WaypointName (NonEmpty Location)
 
-data Navigation = Navigation
-  { waypoints :: M.Map WaypointName (NonEmpty Location)
+data AnnotatedDestination a = AnnotatedDestination
+  { enforceConsistency :: Bool
+  , cosmoLocation :: Cosmic a
```
Comment (byorgey) [inline src/Swarm/Game/Scenario/Topography/Navigation/Portal.hs:35]: Should we rename this to `cosmicLocation` as well?

File: src/Swarm/Game/World.hs [hunk 15]
```diff
@@ -214,6 +222,14 @@ newWorld f = World f M.empty M.empty
 emptyWorld :: t -> World t e
 emptyWorld t = newWorld (WF $ const (t, Nothing))
 
+lookupCosmoTerrain ::
```
Comment (byorgey) [inline src/Swarm/Game/World.hs:225]: If we've renamed the type `Cosmo` to `Cosmic` I think we should also rename any functions that have the word `Cosmo` in them.

**Review by byorgey — APPROVED** [review]

**Pull request merged**